akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala

```diff
@@ -0,0 +1,51 @@
+/**
+ * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.typed.javadsl
+
+import scala.annotation.tailrec
+
+import akka.japi.pf.FI.Apply2
+
+import akka.typed
+import akka.typed.{ Behavior, ExtensibleBehavior, Signal }
+
+case class BehaviorBuilder[T](
+  messageHandlers: List[(Class[_ <: T], (ActorContext[T], T) ⇒ Behavior[T])],
+  signalHandlers:  List[(Class[_ <: Signal], (ActorContext[T], Signal) ⇒ Behavior[T])]
+) {
+
+  def build(): Behavior[T] = {
+    new ExtensibleBehavior[T] {
+      override def receiveMessage(ctx: typed.ActorContext[T], msg: T): Behavior[T] = receive[T](ctx, msg, messageHandlers)
+
+      override def receiveSignal(ctx: typed.ActorContext[T], msg: Signal): Behavior[T] = receive[Signal](ctx, msg, signalHandlers)
+
+      @tailrec
+      private def receive[M](ctx: ActorContext[T], msg: M, handlers: List[(Class[_ <: M], (ActorContext[T], M) ⇒ Behavior[T])]): Behavior[T] =
+        handlers match {
+          case Nil ⇒
+            // emulate scala match error
+            throw new MatchError(msg)
+          case (cls, impl) :: tail ⇒
```

Comment:
What would be a good way to check? I looked at the bytecode:
```
  private <M> akka.typed.Behavior<T> receive(akka.typed.javadsl.ActorContext<T>, M, scala.collection.immutable.List<scala.Tuple2<java.lang.Class<? extends M>, scala.Function2<akka.typed.javadsl.ActorContext<T>, M, akka.typed.Behavior<T>>>>);
    Code:
       0: aload_3
       1: astore        5
       3: getstatic     #43                 // Field scala/collection/immutable/Nil$.MODULE$:Lscala/collection/immutable/Nil$;
       6: aload         5
       8: invokevirtual #49                 // Method java/lang/Object.equals:(Ljava/lang/Object;)Z
      11: ifeq          23
      14: new           #51                 // class scala/MatchError
      17: dup
      18: aload_2
      19: invokespecial #55                 // Method scala/MatchError."<init>":(Ljava/lang/Object;)V
      22: athrow
      23: aload         5
      25: instanceof    #57                 // class scala/collection/immutable/$colon$colon
      28: ifeq          119
      31: aload         5
      33: checkcast     #57                 // class scala/collection/immutable/$colon$colon
      36: astore        7
      38: aload         7
      40: invokevirtual #61                 // Method scala/collection/immutable/$colon$colon.head:()Ljava/lang/Object;
      43: checkcast     #63                 // class scala/Tuple2
      46: astore        8
      48: aload         7
      50: invokevirtual #66                 // Method scala/collection/immutable/$colon$colon.tl$1:()Lscala/collection/immutable/List;
      53: astore        9
      55: aload         8
      57: ifnull        119
      60: aload         8
      62: invokevirtual #69                 // Method scala/Tuple2._1:()Ljava/lang/Object;
      65: checkcast     #71                 // class java/lang/Class
      68: astore        10
      70: aload         8
      72: invokevirtual #74                 // Method scala/Tuple2._2:()Ljava/lang/Object;
      75: checkcast     #76                 // class scala/Function2
      78: astore        11
      80: aload         10
      82: aload_2
      83: invokevirtual #80                 // Method java/lang/Object.getClass:()Ljava/lang/Class;
      86: invokevirtual #84                 // Method java/lang/Class.isAssignableFrom:(Ljava/lang/Class;)Z
      89: ifeq          109
      92: aload         11
      94: aload_1
      95: aload_2
      96: invokeinterface #88,  3           // InterfaceMethod scala/Function2.apply:(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
     101: checkcast     #90                 // class akka/typed/Behavior
     104: astore        6
     106: aload         6
     108: areturn
     109: aload_1
     110: aload_2
     111: aload         9
     113: astore_3
     114: astore_2
     115: astore_1
     116: goto          0
     119: new           #51                 // class scala/MatchError
     122: dup
     123: aload         5
     125: invokespecial #55                 // Method scala/MatchError."<init>":(Ljava/lang/Object;)V
     128: athrow
```
Looks like (40-82) no new `Tuple` is allocated (though does show we might want to move `case Nil` down and replace it with `case _`)